### PR TITLE
Update types for Web Audio API

### PR DESCRIPTION
**Pull request by fand**
Some type definitiions are outdated.

I referred to W3C Editor's Draft 04 August 2017.
https://webaudio.github.io/web-audio-api/

### CODE REVIEW

**Comment by nmote** [comment]
Is this still a work in progress?

**Comment by nmote** [comment]
Could you link to official docs (like MDN) that I could use to verify that this change is correct?

File: lib/bom.js [hunk 1]
```diff
@@ -682,14 +682,15 @@ type PositionOptions = {
...
+
+declare class BaseAudioContext {
   currentTime: number;
   destination: AudioDestinationNode;
   listener: AudioListener;
   sampleRate: number;
-  state: any;
+  state: AudioContextState;
   onstatechange: (ev: any) => any;
-  close(): void;
```
Comment (fand) [inline lib/bom.js]: `BaseAudioContext` doesn't have `close()` in the spec
https://webaudio.github.io/web-audio-api/#BaseAudioContext

File: lib/bom.js [hunk 2]
```diff
@@ -682,14 +682,15 @@ type PositionOptions = {
...
 
-declare class AudioContext {
+type AudioContextState = 'suspended' | 'running' | 'closed';
+
+declare class BaseAudioContext {
   currentTime: number;
   destination: AudioDestinationNode;
   listener: AudioListener;
   sampleRate: number;
-  state: any;
```
Comment (fand) [inline lib/bom.js]: `state` is one of `suspended`, `running` and `closed`.
https://webaudio.github.io/web-audio-api/#enumdef-audiocontextstate

File: lib/bom.js [hunk 3]
```diff
@@ -874,7 +901,7 @@ declare class BiquadFilterNode extends AudioNode {
   Q: AudioParam;
   gain: AudioParam;
   type: 'lowpass'|'highpass'|'bandpass'|'lowshelf'|'highshelf'|'peaking'|'notch'|'allpass';
-  getFrequencyResponse(frequencyHz: Float32Array, magResponse: Float32Array, phaseResponse: Float32Array): BiquadFilterNode;
```
Comment (fand) [inline lib/bom.js]: `getFrequencyResponse` should return `void`
https://webaudio.github.io/web-audio-api/#dom-biquadfilternode-getfrequencyresponse

File: lib/bom.js [hunk 4]
```diff
@@ -901,6 +928,10 @@ declare class GainNode extends AudioNode {
   gain: AudioParam;
 }
 
+declare class IIRFilterNode extends AudioNode {
+  getFrequencyResponse(frequencyHz: Float32Array, magResponse: Float32Array, phaseResponse: Float32Array): void;
+}
+
```
Comment (fand) [inline lib/bom.js]: `IIRFilterNode` spec: 
https://webaudio.github.io/web-audio-api/#dom-biquadfilternode-getfrequencyresponse

File: lib/bom.js [hunk 5]
```diff
@@ -846,6 +872,7 @@ declare class MediaStreamTrackEvent extends Event {
 
 declare class MediaElementAudioSourceNode extends AudioNode {}
 declare class MediaStreamAudioSourceNode extends AudioNode {}
+declare class MediaStreamTrackAudioSourceNode extends AudioNode {}
```
Comment (fand) [inline lib/bom.js]: https://webaudio.github.io/web-audio-api/#mediastreamtrackaudiosourcenode

File: lib/bom.js [hunk 6]
```diff
@@ -714,7 +717,30 @@ declare class AudioContext {
   decodeAudioData(arrayBuffer: ArrayBuffer, decodeSuccessCallback: Function, decodeErrorCallback: Function): void;
   decodeAudioData(arrayBuffer: ArrayBuffer): Promise<AudioBuffer>;
   resume(): Promise<void>;
+}
+
+declare class AudioTimestamp {
+  contextTime: number;
+  performanceTime: number;
+}
```
Comment (fand) [inline lib/bom.js]: `AudioTImestamp` is defined here:
https://webaudio.github.io/web-audio-api/#dictdef-audiotimestamp

File: lib/bom.js [hunk 7]
```diff
@@ -714,7 +717,30 @@ declare class AudioContext {
...
+  createMediaStreamTrackSource(myMediaStreamTrack: MediaStreamTrack): MediaStreamTrackAudioSourceNode;
+  createMediaStreamDestination(): MediaStreamAudioDestinationNode;
+}
+
+declare class OfflineAudioContext extends BaseAudioContext {
+  startRendering(): Promise<AudioBuffer>;
+  suspend(): Promise<void>;
+  length: number;
+  oncomplete: (ev: any) => any;
 }
```
Comment (fand) [inline lib/bom.js]: `OfflineAudioContext` is defined here
https://webaudio.github.io/web-audio-api/#offlineaudiocontext

File: lib/bom.js [hunk 8]
```diff
@@ -714,7 +717,30 @@ declare class AudioContext {
...
+  baseLatency: number;
+  outputLatency: number;
+  getOutputTimestamp(): AudioTimestamp;
   suspend(): Promise<void>;
+  close(): Promise<void>;
+  createMediaElementSource(myMediaElement: HTMLMediaElement): MediaElementAudioSourceNode;
+  createMediaStreamSource(myMediaStream: MediaStream): MediaStreamAudioSourceNode;
+  createMediaStreamTrackSource(myMediaStreamTrack: MediaStreamTrack): MediaStreamTrackAudioSourceNode;
+  createMediaStreamDestination(): MediaStreamAudioDestinationNode;
+}
```
Comment (fand) [inline lib/bom.js]: https://webaudio.github.io/web-audio-api/#audiocontext

**Comment by fand** [comment]
@nmote 
added links to the spec on each diffs 🙏

**Comment by nmote** [comment]
Thanks for the docs. I should be able to find time to look at this tomorrow (Pacific time)

File: lib/bom.js [hunk 9]
```diff
@@ -714,7 +717,30 @@ declare class AudioContext {
   decodeAudioData(arrayBuffer: ArrayBuffer, decodeSuccessCallback: Function, decodeErrorCallback: Function): void;
   decodeAudioData(arrayBuffer: ArrayBuffer): Promise<AudioBuffer>;
   resume(): Promise<void>;
```
Comment (nmote) [inline lib/bom.js]: It looks like `resume` is actually defined on `AudioContext`

File: lib/bom.js [hunk 10]
```diff
@@ -714,7 +717,30 @@ declare class AudioContext {
...
+  close(): Promise<void>;
+  createMediaElementSource(myMediaElement: HTMLMediaElement): MediaElementAudioSourceNode;
+  createMediaStreamSource(myMediaStream: MediaStream): MediaStreamAudioSourceNode;
+  createMediaStreamTrackSource(myMediaStreamTrack: MediaStreamTrack): MediaStreamTrackAudioSourceNode;
+  createMediaStreamDestination(): MediaStreamAudioDestinationNode;
+}
+
+declare class OfflineAudioContext extends BaseAudioContext {
+  startRendering(): Promise<AudioBuffer>;
+  suspend(): Promise<void>;
```
Comment (nmote) [inline lib/bom.js]: `suspend` has a required argument

File: lib/bom.js (hunk 9 above)
Comment (fand) [reply lib/bom.js]: ah I overlooked it. removed on ` [4e32402](/facebook/flow/pull/4570/commits/4e3240273a399ba2e72e0f7362299fb7ad00ba3b)`

File: lib/bom.js (hunk 10 above)
Comment (fand) [reply lib/bom.js]: added on d61bdcd

**Comment by fand** [comment]
@nmote Thank you for review! I fixed the code you pointed out

**Comment by nmote** [comment]
Looks like you removed `resume` from `BaseAudioContext` but didn't add it to `AudioContext`

**Comment by fand** [comment]
😫 added `resume` to AudioContext...

**Pull request closed without merging**